File: pkg/java/src/main/java/dev/openfga/language/DslToJsonTransformer.java
package dev.openfga.language;

import static java.util.stream.Collectors.joining;

import dev.openfga.language.antlr.OpenFGALexer;
import dev.openfga.language.antlr.OpenFGAParser;
import dev.openfga.language.errors.DslErrorsException;
import dev.openfga.language.errors.SyntaxError;
import dev.openfga.sdk.api.model.AuthorizationModel;
import dev.openfga.sdk.api.model.TypeDefinition;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.tree.ParseTreeWalker;

public class DslToJsonTransformer {
    public String transform(String dsl) throws IOException, DslErrorsException {
        return JSON.stringify(parseAuthorisationModel(dsl));
    }

    private AuthorizationModel parseAuthorisationModel(String dsl) throws DslErrorsException {
        var result = parseDsl(dsl);
        if (result.IsFailure()) {
            throw new DslErrorsException(result.getErrors());
        }

        return result.getAuthorizationModel();
    }

    private static final Pattern SPACES_LINE_PATTERN = Pattern.compile("^\\s*$");
    private static final Pattern COMMENTED_LINE_PATTERN = Pattern.compile("^\\s*#.*$");

    private String cleanLine(String line) {
        if (SPACES_LINE_PATTERN.matcher(line).matches()
                || COMMENTED_LINE_PATTERN.matcher(line).matches()) {
            return "";
        }
        var cleanedLine = line.split(Pattern.quote(" #"))[0];
        return cleanedLine.stripTrailing();
    }

    public Result parseDsl(String dsl) {
        var cleanedDsl = Arrays.stream(dsl.split("\n")).map(this::cleanLine).collect(joining("\n"));

        var antlrStream = CharStreams.fromString(cleanedDsl);
        var errorListener = new OpenFgaDslErrorListener();

        var lexer = new OpenFGALexer(antlrStream);
        lexer.removeErrorListeners();
        lexer.addErrorListener(errorListener);
        var tokenStream = new CommonTokenStream(lexer);

        var parser = new OpenFGAParser(tokenStream);
        parser.removeErrorListeners();
        parser.addErrorListener(errorListener);

        var listener = new OpenFgaDslListener(parser);
        new ParseTreeWalker().walk(listener, parser.main());

        return new Result(listener.getAuthorizationModel(), errorListener.getErrors());
    }

    public ModularResult parseModularDsl(String dsl) {
        var cleanedDsl = Arrays.stream(dsl.split("\n")).map(this::cleanLine).collect(joining("\n"));

        var antlrStream = CharStreams.fromString(cleanedDsl);
        var errorListener = new OpenFgaDslErrorListener();

        var lexer = new OpenFGALexer(antlrStream);
        lexer.removeErrorListeners();
        lexer.addErrorListener(errorListener);
        var tokenStream = new CommonTokenStream(lexer);

        var parser = new OpenFGAParser(tokenStream);
        parser.removeErrorListeners();
        parser.addErrorListener(errorListener);

        var listener = new OpenFgaDslListener(parser);
        new ParseTreeWalker().walk(listener, parser.main());

        return new ModularResult(
                listener.getAuthorizationModel(), errorListener.getErrors(), listener.getTypeDefExtensions());
    }

    public static class Result {
        private final AuthorizationModel authorizationModel;
        private final List<SyntaxError> errors;

        public Result(AuthorizationModel authorizationModel, List<SyntaxError> errors) {
            this.authorizationModel = authorizationModel;
            this.errors = errors;
        }

        public AuthorizationModel getAuthorizationModel() {
            return authorizationModel;
        }

        public List<SyntaxError> getErrors() {
            return errors;
        }

        public boolean IsSuccess() {
            return errors.isEmpty();
        }

        public boolean IsFailure() {
            return !IsSuccess();
        }
    }

    public static final class ModularResult extends Result {

        private final Map<String, TypeDefinition> extendedTypeDefinitions;

        public ModularResult(
                AuthorizationModel authorizationModel,
                List<SyntaxError> errors,
                Map<String, TypeDefinition> extendedTypeDefinitions) {
            super(authorizationModel, errors);
            this.extendedTypeDefinitions = extendedTypeDefinitions;
        }

        public Map<String, TypeDefinition> getExtendedTypeDefinitions() {
            return extendedTypeDefinitions;
        }
    }
}

File: pkg/csharp/Transformers/DslToJsonTransformer.cs
using System.Text.RegularExpressions;
using Antlr4.Runtime;
using Antlr4.Runtime.Tree;
using OpenFga.Sdk.Model;
using OpenFga.Language;
using OpenFgaLanguage.Errors;
using OpenFgaLanguage.Listeners;
using OpenFgaLanguage.Utils;

namespace OpenFgaLanguage.Transformers;

public class DslToJsonTransformer
{
    private static readonly Regex SpacesLinePattern = new(@"^\s*$");
    private static readonly Regex CommentedLinePattern = new(@"^\s*#.*$");

    public string Transform(string dsl)
    {
        return Json.Stringify(ParseAuthorisationModel(dsl));
    }

    private AuthorizationModel ParseAuthorisationModel(string dsl)
    {
        var result = ParseDsl(dsl);
        if (result.IsFailure())
        {
            throw new DslErrorsException(result.Errors);
        }

        return result.AuthorizationModel;
    }

    private string CleanLine(string line)
    {
        if (SpacesLinePattern.IsMatch(line) || CommentedLinePattern.IsMatch(line))
        {
            return string.Empty;
        }
        
        var cleanedLine = line.Split(" #")[0];
        return cleanedLine.TrimEnd();
    }

    public Result ParseDsl(string dsl)
    {
        var cleanedDsl = string.Join("\n", dsl.Split('\n').Select(CleanLine));
        
        var inputStream = new AntlrInputStream(cleanedDsl);
        var errorListener = new OpenFgaDslErrorListener();

        var lexer = new OpenFGALexer(inputStream);
        lexer.RemoveErrorListeners();
        lexer.AddErrorListener(errorListener);
        var tokenStream = new CommonTokenStream(lexer);

        var parser = new OpenFGAParser(tokenStream);
        parser.RemoveErrorListeners();
        parser.AddErrorListener(errorListener);

        var listener = new OpenFgaDslListener(parser);
        ParseTreeWalker.Default.Walk(listener, parser.main());

        return new Result(listener.GetAuthorizationModel(), errorListener.Errors.ToList());
    }

    public class Result
    {
        public AuthorizationModel AuthorizationModel { get; }
        public List<SyntaxError> Errors { get; }

        public Result(AuthorizationModel authorizationModel, List<SyntaxError> errors)
        {
            AuthorizationModel = authorizationModel;
            Errors = errors;
        }

        public bool IsSuccess()
        {
            return Errors.Count == 0;
        }

        public bool IsFailure()
        {
            return !IsSuccess();
        }
    }
}
